include/device.h

```diff
@@ -146,32 +138,32 @@ extern "C" {
 	DEVICE_AND_API_INIT(dev_name, drv_name, init_fn, data, cfg_info, \
 			    level, prio, api)
 #else
-#define DEVICE_DEFINE(dev_name, drv_name, init_fn, pm_control_fn,	  \
-		      data, cfg_info, level, prio, api)			  \
-	static struct device_pm _CONCAT(__pm_, dev_name) __used           \
-							= {               \
-		.usage = ATOMIC_INIT(0),                                  \
-		.lock = Z_SEM_INITIALIZER(                               \
-				_CONCAT(__pm_, dev_name).lock, 1, 1),     \
-		.signal = K_POLL_SIGNAL_INITIALIZER(                      \
-				_CONCAT(__pm_, dev_name).signal),         \
-		.event = K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,     \
-				K_POLL_MODE_NOTIFY_ONLY,                  \
-				&_CONCAT(__pm_, dev_name).signal),        \
-	};								  \
-	static struct device_config _CONCAT(__config_, dev_name) __used	  \
-	__attribute__((__section__(".devconfig.init"))) = {		  \
-		.name = drv_name, .init = (init_fn),			  \
-		.device_pm_control = (pm_control_fn),			  \
-		.pm  = &_CONCAT(__pm_, dev_name),                         \
-		.config_info = (cfg_info)				  \
-	};								  \
-	static Z_DECL_ALIGN(struct device) _CONCAT(__device_, dev_name) __used \
-	__attribute__((__section__(".init_" #level STRINGIFY(prio)))) = { \
-		.config = &_CONCAT(__config_, dev_name),		  \
-		.driver_api = api,					  \
-		.driver_data = data,					  \
-	}
+#define DEVICE_DEFINE(dev_name, drv_name, init_fn, pm_control_fn,	\
```

Comment:
done this way it adds code duplication and reduces readability. What about something like this: 
https://gist.github.com/nordic-krch/d299a388124f757c7933f0911018eece

It is using `IF_ENABLED(flag, (code)) macro to inject code only is flag is enabled. It simplifies that code and in the example i tried did not change ram usage.

Reply:
Well, that was not the point of this PR to change the macros and how they are declared. Your proposal sounds totally relevant to me, could you post a PR?